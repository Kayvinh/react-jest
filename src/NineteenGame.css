body {
  background-color: #355e3b;
}

section {
  border: solid 1px white;
  border-radius: 10px;
}

p {
  width: 650px;
  margin: 20px;
  font-size: 200%;
  text-align: center;
  color: yellow;
}

button {
  font-size: 100%;
  padding: 5px 25px;
  border-radius: 10px;
}