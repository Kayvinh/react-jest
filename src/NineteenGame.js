import React, { useState } from "react";
import "./NineteenGame.css";

import { getRandomCard, getPair, score } from "./nineteen";
import Hand from "./Hand";


/** Play a game of Nineteen:
 *
 * (for this simple version, we don't use a deck --- each card is random)
 *
 * State:
 * - playerHand: array of cards
 * - dealerHand: array of cards
 */

function NineteenGame() {
  const [playerHand, setPlayerHand] = useState(getPair);
  const [dealerHand, setDealerHand] = useState([]);
  /** Player draws a card and then dealer plays. */
  function takeCard() {
    setPlayerHand(h => [...h, getRandomCard()]);
    playDealer();
  }
  /** Dealer draws a pair and then hand is scored. */
  function playDealer() {
    setDealerHand(getPair());
  }
  const result = dealerHand.length > 0 ? score(playerHand, dealerHand) : null;

  return (
      <main>
        <Hand cards={dealerHand} />
        <Hand cards={playerHand} />
        {result === null && <p>
          <button id="draw" onClick={takeCard}>Draw</button>
          <button id="pass" onClick={playDealer}>Pass</button>
        </p>}
        {result !== null && <p>{result}</p>}
      </main>
  );
}
// end

export default NineteenGame;
