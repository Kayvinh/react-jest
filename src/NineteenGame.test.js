import * as random from "./random";
import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import NineteenGame from "./NineteenGame.js";

random.choice = jest.fn();

/** Seeds choice fn with data for given cards, eg "AS 7D 4C" */
function _feedChoice(cardsStr) {
  for (let val of cardsStr.replace(/ /g, "")) {
    random.choice.mockReturnValueOnce(val);
  }
}
// end

describe("snapshot tests for stability", function () {
  it("matches initial player hand: 2C 3C", function () {
    _feedChoice("2C 3C");
    const { container } = render(<NineteenGame />);
    expect(container).toMatchSnapshot();
  });
  // end

  it("matches winning game: 2C 3C 4C > 5C 6C", function () {
    _feedChoice("2C 3C 4C   5C 6C");
    const { container } = render(<NineteenGame />);
    fireEvent.click(screen.getByRole("button", { name: "Draw" }));
    expect(container).toMatchSnapshot();
  });
  // end

  it("matches losing game: 2C 3C < 4C 5C", function () {
    _feedChoice("2C 3C   4C 5C");
    const { container } = render(<NineteenGame />);
    fireEvent.click(screen.getByRole("button", { name: "Pass" }));
    expect(container).toMatchSnapshot();
  });
});

// This test does not use the container method
describe("game playing works", function () {
  it("deals to player on draw", function () {
    _feedChoice("2C 3C 4C   5C 6C");
    render(<NineteenGame />);

    const cardElements = screen.getAllByRole("img");
    expect(cardElements).toHaveLength(2);

    fireEvent.click(screen.getByRole("button", { name: "Draw" }));
    let cardsAfterDraw = screen.getAllByRole("img");

    // now player has 3 cards (=9) and dealer has 2 cards (=11)
    expect(cardsAfterDraw).toHaveLength(5);

    const message = screen.getByText("You win!");
    expect(message).toBeTruthy();
  });
  // end

  it("doesn't deal to player on pass", function () {
    _feedChoice("2C 3C   4C 5C");
    render(<NineteenGame />);
    const cardElements = screen.getAllByRole("img");
    expect(cardElements).toHaveLength(2);

    const passBtn = screen.getByRole("button", { name: "Pass" });
    fireEvent.click(passBtn);

    const cardsAfterPass = screen.getAllByRole("img");
    // now player has 2 cards (=5) and dealer has 2 cards (=9)
    expect(cardsAfterPass).toHaveLength(4);

    const message = screen.getByText("You lose!");
    expect(message).toBeTruthy();
  });
});
