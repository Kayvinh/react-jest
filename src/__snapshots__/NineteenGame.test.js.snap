// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for stability matches initial player hand: 2C 3C 1`] = `
<div>
  <main>
    <div
      class="Hand"
    />
    <div
      class="Hand"
    >
      <img
        alt="2C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/2C.png"
      />
      <img
        alt="3C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/3C.png"
      />
    </div>
    <p>
      <button
        id="draw"
      >
        Draw
      </button>
      <button
        id="pass"
      >
        Pass
      </button>
    </p>
  </main>
</div>
`;

exports[`snapshot tests for stability matches losing game: 2C 3C < 4C 5C 1`] = `
<div>
  <main>
    <div
      class="Hand"
    >
      <img
        alt="4C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/4C.png"
      />
      <img
        alt="5C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/5C.png"
      />
    </div>
    <div
      class="Hand"
    >
      <img
        alt="2C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/2C.png"
      />
      <img
        alt="3C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/3C.png"
      />
    </div>
    <p>
      You lose!
    </p>
  </main>
</div>
`;

exports[`snapshot tests for stability matches winning game: 2C 3C 4C > 5C 6C 1`] = `
<div>
  <main>
    <div
      class="Hand"
    >
      <img
        alt="5C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/5C.png"
      />
      <img
        alt="6C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/6C.png"
      />
    </div>
    <div
      class="Hand"
    >
      <img
        alt="2C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/2C.png"
      />
      <img
        alt="3C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/3C.png"
      />
      <img
        alt="4C"
        class="Card"
        src="https://deckofcardsapi.com/static/img/4C.png"
      />
    </div>
    <p>
      You win!
    </p>
  </main>
</div>
`;
